server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,198 @@
+                } catch (Exception e) {
+                    throw new BindTransportException("Failed to re-bind to " + boundSocket.get(), e);
+                }
+
+                return null;
+            }
+
+            final AtomicReference<Exception> lastException = new AtomicReference<>();
+
+            PortsRange portsRange = new PortsRange(PORT.get(environment.settings()));
+            InetAddress localhost = InetAddress.getLoopbackAddress();
+
+            boolean success = portsRange.iterate(portNumber -> {
+                try {
+                    InetSocketAddress socketAddress = new InetSocketAddress(localhost, portNumber);
+                    serverChannel = ServerSocketChannel.open(StandardProtocolFamily.INET);
+                    serverChannel.bind(socketAddress);
+
+                    boundSocket.set(socketAddress);
+                } catch (Exception e) {
+                    lastException.set(e);
+                    return false;
+                }
+                return true;
+            });
+
+            if (success == false) {
+                throw new BindTransportException("Failed to bind to " + NetworkAddress.format(localhost, portsRange), lastException.get());
+            }
+
+            return null;
+        });
+
+        return serverChannel;
+    }
+
+    @Override
+    protected void doStart() {
+        // Mark the service as active, we'll start the listener when ES is ready
+        this.active = true;
+    }
+
+    // package private for testing
+    synchronized void startListener() {
+        if (this.serverChannel != null || this.active == false) {
+            return;
+        }
+
+        this.serverChannel = setupSocket();
+
+        new Thread(() -> {
+            while (serverChannel != null && serverChannel.isOpen()) {
+                AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                    try (SocketChannel channel = serverChannel.accept()) {} catch (IOException e) {
+                        logger.debug("encountered exception while responding to readiness check request", e);
+                    } catch (Exception other) {
+                        logger.warn("encountered unknown exception while responding to readiness check request", other);
+                    }
+                    return null;
+                });
+            }
+        }, "elasticsearch[readiness-service]").start();
+
+        logger.info("readiness service up and running on {}", boundAddress().publishAddress());
+
+        if (WRITE_PORTS_FILE_SETTING.get(environment.settings())) {
+            writePortsFile(environment, "readiness", boundAddress());
+        }
+    }
+
+    @Override
+    protected void doStop() {
+        this.active = false;
+        stopListener();
+    }
+
+    // package private for testing
+    synchronized void stopListener() {
+        try {
+            if (this.serverChannel != null) {
+                this.serverChannel.close();
```

Comment:
Completely closing the socket risks letting someone else grab it while we're unready. Again, maybe no big deal at the moment since we only become unready when shutting down, but maybe not forever. Would it work instead to keep the socket open but only to `accept()` connections on it while we're ready?

Reply:
I considered that, but I was concerned that it wouldn't work well in all cases. Namely, the probe will connect, but since there's no listener to accept the connection, it will just hang and rely on the probe side connect timeout to expire and disconnect.

I suppose there's no major problem here, it will just take much longer to determine that Elasticsearch is down, relying on a client side timeout instead of an immediately rejected connection. 

If this is acceptable, it's better to just end the thread listener loop and keep the port open.

Reply:
I'm not 100% sure about this, would be good to get some other opinions. Timing out is not as nice as an outright rejection indeed. Maybe we don't care enough about someone else pinching our port.

Reply:
I chatted with @ChrisHegarty this morning and we experimented a bit with various options to not run the accept loop. It causes an issue with socket buffers filling up and eventually the ServerSocketChannel gets blocked, causing any connect to timeout until accept runs and clears up the buffers. I think for now I'll stick to the accept thread approach and shutting everything down to avoid problematic scenarios.